pkg/cri/server/container_create_linux.go

```diff
@@ -96,6 +106,8 @@ func (c *criService) containerMounts(sandboxID string, config *runtime.Container
 			HostPath:       c.getResolvPath(sandboxID),
 			Readonly:       securityContext.GetReadonlyRootfs(),
 			SelinuxRelabel: true,
+			UidMappings:    uidMappings,
+			GidMappings:    gidMappings,
```

Comment:
TODO: review if we should add mappings for devShm below too or not

Reply:
I think we can add a TODO for tmpfs, since it requires latest kernel.

REF: https://lore.kernel.org/lkml/20230217080552.1628786-1-brauner@kernel.org/

Reply:
```
2023-07-19T14:49:13.3115747Z === RUN   TestPodUserNS/userns_uid_mapping
2023-07-19T14:49:13.3134727Z     pod_userns_linux_test.go:107: Create a sandbox with userns
2023-07-19T14:49:13.4670368Z     main_test.go:706: Image "ghcr.io/containerd/busybox:1.36" already exists, not pulling.
2023-07-19T14:49:13.4671157Z     pod_userns_linux_test.go:135: Create a container for userns
2023-07-19T14:49:15.3180985Z     pod_userns_linux_test.go:144: Start the container
2023-07-19T14:49:15.3749009Z E0719 14:49:15.374332   84106 remote_runtime.go:270] StartContainer "3fd3ff60e0c5a9e3ced64ab211ed859549f5f75c034633e7692b561e73f41fef" from runtime service failed: rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: can't get final child's PID from pipe: EOF: unknown
2023-07-19T14:49:15.3749920Z     pod_userns_linux_test.go:145: 
2023-07-19T14:49:15.3750601Z         	Error Trace:	/home/runner/work/containerd/containerd/integration/pod_userns_linux_test.go:145
2023-07-19T14:49:15.3751152Z         	Error:      	Received unexpected error:
2023-07-19T14:49:15.3753229Z         	            	rpc error: code = Unknown desc = failed to create containerd task: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: can't get final child's PID from pipe: EOF: unknown
2023-07-19T14:49:15.3753922Z         	Test:       	TestPodUserNS/userns_uid_mapping
```

The CI fails because the kernel v5.15 (ubuntu 22.04) doesn't support tmpfs idmapped.

Reply:
Yes, we merged this in 6.3. And dev/shm is exposed empty (no prepopulated files there), so we can avoid the idmap mount there, and just lower the kernel requirements for our CI.


btw, the error is not clear because of this issue (not sure if it is on runc or containerd): https://github.com/containerd/containerd/issues/8144. If you just run runc with that config.json, a previous line to the one shown here says that the filesystem lacks kernel support for idmap mounts.

I don't know how you realized what the error meant, I admire your debugging skills :-)

Reply:
Sorry for the late reply. I have azure vm as devbox which is the same to github action vm. They are using kernel v5.15.x (ubuntu 22.04). It's easy to debug. I read your comment about tmpfs so that I realized that the error might be caused by kernel. So, I delete the code about devShm and the test pass.

> btw, the error is not clear because of this issue (not sure if it is on runc or containerd)

After I confirmed that it's caused by kernel syscall, I added the following code to confirm it again.

```diff
diff --git a/runtime/v2/runc/container.go b/runtime/v2/runc/container.go
index 448338561..b426f3dfc 100644
--- a/runtime/v2/runc/container.go
+++ b/runtime/v2/runc/container.go
@@ -133,6 +133,8 @@ func NewContainer(ctx context.Context, platform stdio.Platform, r *task.CreateTa
                return nil, errdefs.ToGRPC(err)
        }
        if err := p.Create(ctx, config); err != nil {
+               data, werr := os.ReadFile(r.Bundle + "/log.json")
+               log.G(ctx).WithError(werr).Warn(string(data))
                return nil, errdefs.ToGRPC(err)
        }
        container := &Container{
diff --git a/script/test/cri-integration.sh b/script/test/cri-integration.sh
index 442054a4b..f92612f4f 100755
--- a/script/test/cri-integration.sh
+++ b/script/test/cri-integration.sh
@@ -17,6 +17,7 @@
 set -o errexit
 set -o nounset
 set -o pipefail
+set -x

 basedir="$(dirname "${BASH_SOURCE[0]}")"
 source "${basedir}/utils.sh"
diff --git a/vendor/github.com/containerd/go-runc/runc.go b/vendor/github.com/containerd/go-runc/runc.go
index 61646df6a..c8eb2ee74 100644
--- a/vendor/github.com/containerd/go-runc/runc.go
+++ b/vendor/github.com/containerd/go-runc/runc.go
@@ -178,7 +178,7 @@ func (r *Runc) startCommand(cmd *exec.Cmd) (chan Exit, error) {

 // Create creates a new container and returns its pid if it was created successfully
 func (r *Runc) Create(context context.Context, id, bundle string, opts *CreateOpts) error {
-       args := []string{"create", "--bundle", bundle}
+       args := []string{"--debug", "create", "--bundle", bundle}
        if opts == nil {
                opts = &CreateOpts{}
        }

```

 I saw that `failed to change mount attributes, maybe the filesystem doesn't supports ID-mapped mounts`. So I confirm that error. I think the best way is to use kprobe to watch the sys_mount_setattr syscall argument and return.

```
level=warning msg="{\"level\":\"debug\",\"msg\":\"nsexec[160153]: =\\u003e nsexec container setup\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec[160153]: set process as non-dumpable\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-0[160153]: ~\\u003e nsexec stage-0\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-0[160153]: spawn stage-1\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-0[160153]: -\\u003e stage-1 synchronisation loop\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-1[160156]: ~\\u003e nsexec stage-1\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-1[160156]: setns(0x10000000) into user namespace (with path /proc/160055/ns/user)\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-1[160156]: setns(0x8000000) into ipc namespace (with path /proc/160055/ns/ipc)\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-1[160156]: setns(0x4000000) into uts namespace (with path /proc/160055/ns/uts)\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-1[160156]: setns(0x40000000) into net namespace (with path /proc/160055/ns/net)\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-1[160156]: setns(0x20000000) into pid namespace (with path /proc/160055/ns/pid)\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-1[160156]: unshare remaining namespaces (except cgroupns)\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-1[160156]: request stage-0 to send idmap sources\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-0[160153]: stage-1 requested to open idmap sources\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-0[160153]: ~\\u003e sending idmap source: /var/lib/containerd-test/io.containerd.grpc.v1.cri/sandboxes/b72a4053baf2add0efec24d7503d45b516b58d97ef9c72af8a19ab5221325e24/hostname with mapping from: /proc/160156/ns/user\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-0[160153]: ~\\u003e sending idmap source: /var/lib/containerd-test/io.containerd.grpc.v1.cri/sandboxes/b72a4053baf2add0efec24d7503d45b516b58d97ef9c72af8a19ab5221325e24/hosts with mapping from: /proc/160156/ns/user\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"debug\",\"msg\":\"nsexec-0[160153]: ~\\u003e sending idmap source: /var/lib/containerd-test/io.containerd.grpc.v1.cri/sandboxes/b72a4053baf2add0efec24d7503d45b516b58d97ef9c72af8a19ab5221325e24/resolv.conf with mapping from: /proc/160156/ns/user\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"level\":\"fatal\",\"msg\":\"nsexec-0[160153]: failed to change mount attributes, maybe the filesystem doesn't supports ID-mapped mounts: Invalid argument\",\"time\":\"2023-07-21T12:46:49Z\"}\n{\"file\":\"github.com/opencontainers/runc/utils.go:62\",\"func\":\"main.fatalWithCode\",\"level\":\"error\",\"msg\":\"runc create failed: unable to start container process: can't get final child's PID from pipe: EOF\",\"time\":\"2023-07-21T12:46:49Z\"}\n" error="<nil>" runtime=io.containerd.runc.v2

```

Reply:
Yes. We should enhance shim-runc log for debug. :)

Reply:
@fuweid oh, great .Thanks a lot for the info!

Are you sure CI is using ubuntu 22.04? It seems it is using 20.04 for me: https://github.com/containerd/containerd/actions/runs/5622982505/job/15236863761?pr=8287#step:1:4.

See also here: https://github.com/containerd/containerd/blob/59959b31ff22fec86841b6f8cec2feec9c2d77b4/.github/workflows/ci.yml#L388-L390

Which is unfortunate, we need to skip all the tests in that old kernel version. How shall I proceed?

Do you know if we should add one more Ubuntu version to the matrix (22.04) or just test against 22.04? Or just skipping the test seems fine for you?

Reply:
https://github.com/containerd/containerd/actions/runs/5622982505/job/15236863335?pr=8287 It seems like the error is not integration, not related to kernel. The 20.04 is also using kernel v5.15.x https://github.com/actions/runner-images/blob/main/images/linux/Ubuntu2004-Readme.md. Could you revisit your code again? I think the error might be introduced in recently.

Reply:
Oh, yes, I had a typo in the new integration tests (was checking if the fs in /run supported idmap mounts, that is tmpfs and was skipping, but I need to test for /var/... because there is where we put the /etc/hostname and other files). Great! :)